plugins/inputs/vault/vault.go

```diff
@@ -0,0 +1,219 @@
+
+  ## Set response_timeout (default 5 seconds)
+  # response_timeout = "5s"
+
+  ## Optional TLS Config
+  # tls_ca = /path/to/cafile
+  # tls_cert = /path/to/certfile
+  # tls_key = /path/to/keyfile
+`
+
+func init() {
+	inputs.Add("vault", func() telegraf.Input {
+		return &Vault{
+			ResponseTimeout: config.Duration(5 * time.Second),
+		}
+	})
+}
+
+// SampleConfig returns a sample config
+func (n *Vault) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns a description of the plugin
+func (n *Vault) Description() string {
+	return "Read metrics from the Vault API"
+}
+
+func (n *Vault) Init() error {
+	if n.URL == "" {
+		n.URL = "http://127.0.0.1:8200"
+	}
+
+	if n.Token == "" && n.TokenString == "" {
+		return fmt.Errorf("token missing")
+	}
+
+	if n.Token != "" && n.TokenString != "" {
+		return fmt.Errorf("both token and token_string are set")
+	}
+
+	if n.Token != "" {
+		token, err := os.ReadFile(n.Token)
+		if err != nil {
+			return fmt.Errorf("reading file failed: %v", err)
+		}
+		n.TokenString = strings.TrimSpace(string(token))
+	}
+
+	tlsCfg, err := n.ClientConfig.TLSConfig()
+	if err != nil {
+		return fmt.Errorf("setting up TLS configuration failed: %v", err)
+	}
+
+	n.roundTripper = &http.Transport{
+		TLSHandshakeTimeout:   5 * time.Second,
+		TLSClientConfig:       tlsCfg,
+		ResponseHeaderTimeout: time.Duration(n.ResponseTimeout),
+	}
+
+	return nil
+}
+
+// Gather, collects metrics from Vault endpoint
+func (n *Vault) Gather(acc telegraf.Accumulator) error {
+	sysMetrics := &SysMetrics{}
+	err := n.loadJSON(n.URL+"/v1/sys/metrics", sysMetrics)
+	if err != nil {
+		return err
+	}
+
+	err = buildVaultMetrics(acc, sysMetrics)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (n *Vault) loadJSON(url string, v interface{}) error {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("X-Vault-Token", n.TokenString)
+	req.Header.Add("Accept", "application/json")
+
+	resp, err := n.roundTripper.RoundTrip(req)
+	if err != nil {
+		return fmt.Errorf("error making HTTP request to %s: %s", url, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("%s returned HTTP status %s", url, resp.Status)
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(v)
+	if err != nil {
+		return fmt.Errorf("error parsing json response: %s", err)
+	}
+
+	return nil
+}
+
+// buildVaultMetrics, it builds all the metrics and adds them to the accumulator
+func buildVaultMetrics(acc telegraf.Accumulator, sysMetrics *SysMetrics) error {
+	t, err := time.Parse(timeLayout, sysMetrics.Timestamp)
+	if err != nil {
+		return fmt.Errorf("error parsing time: %s", err)
+	}
+
+	for _, counters := range sysMetrics.Counters {
+		tags := make(map[string]string)
+		for key, val := range counters.baseInfo.Labels {
+			convertedVal, err := internal.ToString(val)
+			if err != nil {
+				return fmt.Errorf("error converting value: %s", err)
```

Comment:
This error message exists three times in this function, for `counters`, `gauges` and `summary`. When seeing an error you have not enough information to debug the problem. How about
```suggestion
				return fmt.Errorf("converting counter %s=%v failed: %v", key, val, err)
```
and likewise for the other two instances. So given you have the raw data you can (maybe) find and reproduce the issue...